tests/unit/services/events-test.ts

```diff
@@ -1,11 +1,188 @@
 import { module, test } from 'qunit';
 import { setupTest } from 'ember-qunit';
+import { settled } from '@ember/test-helpers';
+import { prefixPath, exactPath } from 'hyperevents/services/events-service';
+
+const eventPayload: any = {
+  resource: '/toto',
+  payload: {
+    foo: 'bar'
+  }
+};
+
+class MockSocket {
+  callbacks: any = {};
+
+  addEventListener(param: any, func: any) {
+    this.callbacks[param] = func;
+  }
+
+  close() {}
+
+  send() {}
+}
 
 module('Unit | Service | events', function (hooks) {
   setupTest(hooks);
 
-  // Replace this with your real tests.
-  test('it exists', function (assert) {
-    assert.true(true)
+  hooks.beforeEach(function () {
+    this.service = this.owner.lookup('service:events-service');
+    this.mockedSocket = new MockSocket();
+    this.service._buildSocket = () => {
+      return this.mockedSocket;
+    };
+    this.service._backoffDelay = () => {
+      return 0;
+    };
+  });
+
+  test('it exists', async function (assert) {
+    assert.ok(this.service);
+  });
+
+  test('When the socket is connected it notifies the "onConnected()" subscribers', async function (assert) {
+    this.connected = false;
+
+    this.service.establishConnection('/fakeurl');
+    await settled();
+
+    this.service.onConnected().subscribe(() => {
+      this.connected = true;
+    });
+
+    this.mockedSocket.callbacks['open']();
+    assert.true(this.connected);
+  });
+
+  test('The connection is retried if the socket connection fails', async function (assert) {
+    this.pass = false;
+    this.connected = false;
+
+    this.service._buildSocket = () => {
+      if (!this.pass) {
+        this.pass = true;
+        throw new Error('fake error');
+      }
+      return this.mockedSocket;
+    };
+
+    this.service.establishConnection('/fakeurl');
```

Comment:
It's not the purpose of the PR, but the `establishConnection` question me a bit.
The design of the `establishConnection` (if I am not wrong)  make the init of the connection async. I'm not sure if this is a good design since currently if the connection fail, you will retry infinitely and the flow will continue like if the connection is open. This could lead to some issue, no ?

Reply:
> The design of the establishConnection (if I am not wrong) make the init of the connection async.

That's correct.

>  will retry infinitely and the flow will continue like if the connection is open. 

One should not assume that the connection is opened after calling `establishConnection`. The expected way of using this is after receiving `onConnected`.

But, even if we can't reach events-server, user code can still subscribe to message events, independently of the connection state, they just won't receive any message until the connection is established, but that's none of their concern (Or if they care, they subscribe to `onConnected` :sweat_smile:).

Does this make any sense?

Reply:
Yes, currently we don't emit anything, so it's not a big deal. Just it's could be useful to known that the connection is closed ?

Not really related, but the `socket.close()` method trigger the `close` event ? (because we try to reconnect in case of close event)

Reply:
> Just it's could be useful to known that the open connection fail ?

Indeed, we could have a `onConnectionAttemptFailed`  or smth like that.

> Not really related, but the socket.close() method trigger the close event ? (because we try to reconnect in case of close event)

We try only if it is an abnormal closure (https://github.com/upfluence/hyperevents/blob/9c8f0f931b56c4f30c34a9b74d64a5719dc3b8ef/addon/services/events-service.ts#L105)